centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
 
-## II Interface
-This section describes the standard interface between the OSML and the ObjectSystem driver itself.
-
-### A.  Initialization
-Each OS Driver will have an initialization function, normally named xxxInitialize() where 'xxx' is the driver's abbreviative prefix.  This prefix should be attached to each and every function within the OS driver for consistency and project management.  Normally 'xxx' is two to four characters, all lowercase.  This initialization function is called when the Centrallix starts up, and at least at the present time, this initial call to the OS driver must be manually added to the appropriate startup code, currently found in 'centrallix.c'.
-
-Within the initialization function, the driver should initialize all necessary global variables and register itself with the OSML.  Global variables should all be placed inside a single global 'struct', which is normally named similarly to the driver's prefix, except normally in all uppercase.  Under no circumstances should global variables be accessed outside of the module, except via the module's functions.
+Using the example above, we can query from the database using a statement like `select :title from /OMSS_DB/JNetHelp/rows`, which will open a sybase driver instance, then open a query and repeatedly fetch rows, getting the `title` attribute from each row.
 
-To register with the OSML, the driver must first allocate an ObjDriver structure and fill in its contents.  
 
-    pObjDriver drv;
-
-    drv = (pObjDriver)nmMalloc(sizeof(ObjDriver));
+## II Interface
+This section describes the standard interface between the OSML and the ObjectSystem driver itself.  Every driver should implement certain required functions.  (**Note**: Many drivers "implement" some required functions to simply fail with a not implemented or not supported error.  For example, most database drivers "implement" `Read()` and `Write()` this way because database content should be queried, not read).  Various optional functions are also available, which a driver is not required to implement.
+
+<!-- TODO: Greg
+  --- Double check the information in this table. Some of it is missing,
+  --- and I had to guess a lot of it by looking at how various drivers were
+  --- implemented. Also, which functions are optional and which are required
+  --- seems very chaotic and I vaguely remember there being a driver that
+  --- does not implement a ton of "required" functions, so please double
+  --- check that this information is 100% correct.
+  --->
+The driver should implement an `Initialize()` function, as well as the following (* indicates required functions):
+| Function Name                                             | Description
+| --------------------------------------------------------- | ------------
+| [Open](#function-open)*                                   | Opens a new driver instance object on a given root node.
+| [OpenChild](#function-openchild)                          | ???
```

Comment:
See discussion in the OpenChild section below.  Not required.

Reply:
Updated